File: Algorithms/Python/0006/0006.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s
        res = ""
        n, offset = len(s), 2 * (numRows - 1)
        res = res + s[0 : n : offset]
        for row in range(1, numRows - 1):
            for i in range(row, n, offset):
                res = res + s[i]
                idx = i + 2 * (numRows - 1 - row)
                if idx < n:
                    res = res + s[idx]
        res = res + s[numRows - 1 : n : offset]
        return res

File: Algorithms/C++/0006/0006.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows == 1) {
            return s;
        }
        string res;
        int offset = 2 * (numRows - 1);
        for (int j = 0; j < s.size(); j += offset) {
            res += s[j];
        }
        for (int i = 1; i < numRows - 1; i++) {
            for (int j = i; j < s.size(); j += offset) {
                res += s[j];
                int idx = j +  2 * (numRows - 1 - i);
                if (idx < s.size()) {
                    res += s[idx];
                }
            }
        }
        for (int j = numRows - 1; j < s.size(); j += offset) {
            res += s[j];
        }
        return res;
    }
};
